### PR TITLE
Basic implementation of Clumper.from_jsonl(..)

### PR DESCRIPTION
**Pull request by samarpan-rai**
Related to [this issue](https://github.com/koaning/clumper/issues/16). 

- [x] Basic implementation of importing Clumper.from_jsonl(file_path, lines). You can specify if you want to limit the number of lines
- [x] Detailed documentation on the function 
- ~~[ ]  Add example image once the function is complete~~ I don't need to do this
- [x] Auto download file if provided as URL 
- ~~[ ] Add further tests to catch cases where parsing is not handled well by `jsonlines` library~~ Not using  `jsonlines` library.  Using  `open` function to read file line by line and use `json` library to parse the line to JSON string.
- [x] Auto downloads file specified with URL and read it line by line if specified else reads in the entire file. 
- [ ] Refactoring and simplification

### CODE REVIEW

**Comment by koaning** [comment]
Let's keep the conversation about the PR here. That said; 

I usually keep a `data` folder in the `tests` directory for this sort of thing. Feel free to create that and add a small file there. Since you've got a PR ready though, we should keep this conversation to the PR. Github issues are usually meant to discuss what features should go into a ticket. Implementation discussions should be had on the PR side.

File: tests/test_read_write/test_read_jsonl.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    path = str(datafiles)  # Convert from py.path object to path (str)
+
+    file_path = str(os.path.join(path, "cards.jsonl"))
+    assert os.path.isfile(file_path)  # Make sure its a file
+    assert (
+        len(Clumper.read_jsonl(file_path)) == 4
+    )  # There are 4 rows in the sample file
+
+
+@pytest.mark.datafiles("./tests/test_read_write/sample_jsonl_files/cards.jsonl")
```
Comment (koaning) [inline tests/test_read_write/test_read_jsonl.py:17]: Might be nice to add a `pytest.mark.parametrize` here so we can check that it works for 0 lines, 1 line, 2 lines, etc.

File: setup.py [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 
+
 base_packages = []
 
-test_packages = [
+test_packages = base_packages + [
     "pytest>=5.4.3",
     "black>=19.10b0",
     "flake8>=3.8.3",
+    "pytest-datafiles>=2.0",
```
Comment (koaning) [inline setup.py:10]: Is there a reason why you need this? Reading in a file could also be done with a `pytest.fixture` here. If we can omit dependencies that would be preferable.

File: clumper/clump.py [hunk 3]
```diff
@@ -62,6 +62,33 @@ def read_json(cls, path):
...
+        Reads in a jsonl file. You can also specify how many lines you want to read in.
+
+        """
+        data_array = []
+        # Case 1 : Handle local files
+        if path.startswith("https:") and path.startswith("http:"):
+            raise ValueError(
+                "Currently the jsonl files must be in the local machine. Please specify a local file."
+            )
+        else:
```
Comment (koaning) [inline clumper/clump.py:77]: Do you need the `else` statement here?

File: tests/test_read_write/test_read_jsonl.py (hunk 1 above)
Comment (samarpan-rai) [reply tests/test_read_write/test_read_jsonl.py:17]: Good idea!

File: setup.py (hunk 2 above)
Comment (samarpan-rai) [reply setup.py:10]: You're right. I could instead create a function that returns the name of the file

File: clumper/clump.py (hunk 3 above)
Comment (samarpan-rai) [reply clumper/clump.py:77]: I guess not since I am raising an error. Once I do implement the code to automatically download the file then I might need an else statement but that's just a speculation. For now I will remove it.

**Comment by samarpan-rai** [comment]
I have updated my code with your comments. Shall I open a new PR or will the same PR be able to access my latest master branch commits? Sorry if that's a noob question. I am quite new to git PR in real life.

**Comment by samarpan-rai** [comment]
Related to [this issue](https://github.com/koaning/clumper/issues/16). 

- [x] Basic implementation of importing Clumper.from_jsonl(file_path, lines). You can specify if you want to limit number of lines
- [x] Detailed documentation on the function 
- [ ]  Add example image once the function is complete
- [ ] Auto download file if provided as url 
- ~~[ ] Add further tests to catch cases where parsing is not handled well by `jsonlines` library~~ Not using  `jsonlines` library.  Using  `open` function to read file line by line and use `json` library to parse the line to JSON string.

**Comment by koaning** [comment]
Also. Don't worry about the documentation image. It is best if I do it in order to maintain the same handwriting in the entire project.

**Comment by samarpan-rai** [comment]
Alright. Will you upload a sample jsonl files for the test and share the link?

**Comment by koaning** [comment]
This should work: https://calmcode.io/datasets/pokemon.jsonl.

**Comment by samarpan-rai** [comment]
I am currently using `requests` package to stream files line by line such that if the user provides a really large file, it doesn't have to download it completely but only up to a certain line. I think it should be ready for the basic functionalities. Please give me feedback on how I could improve.

**Comment by koaning** [comment]
I think I would prefer to omit `requests`. I think you've done a like-able implementation here but we currently have a library with zero dependencies and that is actually quite a feature. One I'd prefer to keep.

If that means that we need to download an entire file before we filter the lines out ... I can live with that. Feel free to challenge me on this ... but if you've got a huge file you want to analyse ... better to pull it in first. Also, performance isn't the main goal of the library. I suppose simplicity is more important. 

Again, feel free to challenge me on this.

**Comment by samarpan-rai** [comment]
> I think I would prefer to omit `requests`. I think you've done a like-able implementation here but we currently have a library with zero dependencies and that is actually quite a feature. One I'd prefer to keep.
> 
> If that means that we need to download an entire file before we filter the lines out ... I can live with that. Feel free to challenge me on this ... but if you've got a huge file you want to analyse ... better to pull it in first. Also, performance isn't the main goal of the library. I suppose simplicity is more important.
> 
> Again, feel free to challenge me on this.

In terms of simplicity, you are right but in terms of playing with large files, it will be problematic if a user wants to load in a few 100 lines and just do a quick analysis. If the goal of this package is simplicity, then what you prefer is right. I have implemented the version that downloads the file.

File: tests/conftest.py [hunk 4]
```diff
@@ -9,3 +9,13 @@ def base_clumper():
         for i, c in enumerate("abcdefghijklmnopqrstuvwxyz")
     ]
     return Clumper(data)
+
+
+@pytest.fixture(scope="module")
+def single_local_jsonl_file_path():
```
Comment (koaning) [inline tests/conftest.py:15]: Adding a fixture for a single file path that is only used in a single test is a bit of overkill. If multiple tests in a single file use it then it makes sense to add a fixture for a file level. But in this case we're dealing with fixtures that are used by one test in one file. Best to remove the fixtures here and just use the path directly in the test.

File: tests/test_read_write/test_read_jsonl.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import pytest
+from clumper import Clumper
+
+
+@pytest.mark.parametrize("lines, expected", [(0, 0), (1, 1), (2, 2)])
```
Comment (koaning) [inline tests/test_read_write/test_read_jsonl.py:5]: Probably good to also add a case `(None, 4)`. I'm wondering if it also makes sense to add a case `(5, 4)`. To me it feels like appropriate behavior to read in all the lines if `lines` > number of items.

Comment (koaning) [reply tests/test_read_write/test_read_jsonl.py:5]: You can also do something similar for the test below.

File: tests/conftest.py (hunk 4 above)
Comment (koaning) [reply tests/conftest.py:15]: Same for the fixture below.

File: setup.py [hunk 6]
```diff
@@ -1,8 +1,9 @@
 from setuptools import setup, find_packages
 
+
 base_packages = []
```
Comment (koaning) [inline setup.py:4]: `base_packages` as a variable can be removed now.

File: clumper/clump.py [hunk 7]
```diff
@@ -62,6 +62,42 @@ def read_json(cls, path):
...
+        Reads in a jsonl file. You can also specify how many lines you want to read in.
+
+        """
+        data_array = []
+
+        if lines is not None:
+            assert lines >= 0, "Number of lines to read must be non-negative"
+
+        def iterate_over_file(file_handler):
+            nonlocal data_array, lines
```
Comment (koaning) [inline clumper/clump.py:77]: Do you really need to refer to the `data_array` global here?

Comment (koaning) [reply clumper/clump.py:77]: Can't we just pass `data_array` and `lines` in here as inputs?

File: clumper/clump.py [hunk 8]
```diff
@@ -62,6 +62,42 @@ def read_json(cls, path):
...
+        try:
+            # Case 1 : Open cloud file in stream
+            if path.startswith("https:") or path.startswith("http:"):
+                with urllib.request.urlopen(path) as f:
+                    iterate_over_file(f)
+                    return Clumper(data_array)
+            # Case 2 : Local file
+            with open(path) as f:
+                iterate_over_file(f)
+                return Clumper(data_array)
```
Comment (koaning) [inline clumper/clump.py:95]: Potential simplification possible. I notice two blocks of; 

```python
iterate_over_file(f)
return Clumper(data_array)
```

Can't we use the if statment to decided how to grab the file reference `f` but from there apply the same operation?

File: clumper/clump.py [hunk 9]
```diff
@@ -62,6 +62,42 @@ def read_json(cls, path):
...
+
+        except Exception:
+            print("Error occured during parsing jsonl file")
+            raise
```
Comment (koaning) [inline clumper/clump.py:99]: Why not;

```python
raise RunTimeError("Error occured during parsing jsonl file")
```

This is typically a lot safer because then other users of our library can catch the error as well as the error message. Might also be nice to write a test confirming that this error is actually being raised.

File: clumper/clump.py (hunk 8 above)
Comment (koaning) [reply clumper/clump.py:95]: This way, you may not need the `iterate_over_file` anymore either.

**Comment by koaning** [comment]
Side question. I might want to add some calmcode videos on reviewing and collaboration on github. Would you be opposed if I used our interactions in some examples? Feel free to say no if you don't prefer it.

File: tests/test_read_write/test_read_jsonl.py (hunk 5 above)
Comment (samarpan-rai) [reply tests/test_read_write/test_read_jsonl.py:5]: > Side question. I might want to add some calmcode videos on reviewing and collaboration on GitHub. Would you be opposed if I used our interactions in some examples? Feel free to say no if you don't prefer it.

No problem. I have learnt so much from your videos at calmcode so I am happy to give something back to the community, even if that means they see my mistakes.

File: tests/conftest.py (hunk 4 above)
Comment (samarpan-rai) [reply tests/conftest.py:15]: My logic was: Maybe I want to add more tests in the future, and I don't want to keep on copy-pasting the file path. Do you think that this is premature optimization?

File: clumper/clump.py (hunk 8 above)
Comment (samarpan-rai) [reply clumper/clump.py:95]: Yes, I can do this. Now that I am not using the requests library, I can simplify it.

File: tests/test_read_write/test_read_jsonl.py (hunk 5 above)
Comment (samarpan-rai) [reply tests/test_read_write/test_read_jsonl.py:5]: > Probably good to also add a case `(None, 4)`. I'm wondering if it also makes sense to add a case `(5, 4)`. To me it feels like appropriate behavior to read in all the lines if `lines` > number of items.

Regarding `(None, 4)` case: 
This is handled by existing code. 

Regarding `(5, 4)` case:
My code already handles this as iterate through each line of the file. The iteration ends at the end of the file so `lines` > number of items then the program never reaches the `break` keyword so the entire file will be read.  

Added test case to both local and cloud file.
Maybe another test case could be to read negative lines and expect an error?

File: clumper/clump.py [hunk 10]
```diff
@@ -62,6 +62,56 @@ def read_json(cls, path):
...
+        clump = Clumper.read_jsonl("tests/cards.jsonl")
+        assert len(clump) == 5
+
+
+        clump = Clumper.read_jsonl("https://calmcode.io/datasets/pokemon.jsonl")
+        assert len(clump) == 800
+
+        """
+
+        assert path.lower().endswith(
```
Comment (koaning) [inline clumper/clump.py:84]: I think in the future I might change this to a `ValueError` but that's really a detail.

**Comment by koaning** [comment]
LGTM! Thanks!

File: tests/conftest.py (hunk 4 above)
Comment (koaning) [reply tests/conftest.py:15]: I think it is. But it's a detail that I can easily adres later down the line.

**Comment by samarpan-rai** [comment]
Limitation: The user can read in **.json** file by just renaming it to **.jsonl**. Clumper would still parse it one line as a big dictionary. Unexpected behaviour will happen during analysis.

To reproduce:
1. Rename `pokemon.json` to  `pokemon.jsonl`
2. Read it and load to `Clumper`
```py
from clumper import Clumper
wrongly_parsed = Clumper.read_jsonl("pokemon.jsonl")
```
3. You can see that the `len` returns 1
```py
print(len(wrongly_parsed))
```

I couldn't find how to verify if the file being read is actually JSONL apart from looking at its extension. Any feedback/suggestion to improve this welcome.

**Comment by koaning** [comment]
@samarpan-rai Could you make a seperate issue for this? I can argue this is still appropriate behavior but I'd prefer to separate discussions where possible.
